sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,684 @@
+  The effect is the same regardless of whether redeclaration are so decorated.
+
+* The same function may be decorated with at most one of these macros.
+  However, the same function may be decorated multiple times with the same
+  macro (with the same arguments).
+  Programs that decorate the same function with more than one of these macros
+  or with multiple instances of the same macro with different arguments are ill
+  formed.
+
+* If a function is decorated with one of these macros in one translation unit,
+  that function must also be decorated with the same macro (with the same
+  arguments) in all other translation units.
+
+The following table provides additional details about these macros:
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+#define SYCL_EXT_ONEAPI_RANGE_KERNEL(Dims) /*unspecified*/
+----
+!====
+
+Indicates that the function is a free function kernel that is invoked with a
+simple `range` iteration space of `Dims` dimensions.
+The `Dims` parameter must be an integral constant expression that specifies a
+valid number of dimensions for the `range` class.
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+#define SYCL_EXT_ONEAPI_ND_RANGE_KERNEL(Dims) /*unspecified*/
+----
+!====
+
+Indicates that the function is a free function kernel that is invoked with an
+`nd_range` iteration space of `Dims` dimensions.
+The `Dims` parameter must be an integral constant expression that specifies a
+valid number of dimensions for the `nd_range` class.
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+#define SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL /*unspecified */
+----
+!====
+
+Indicates that the function is a free function kernel that is invoked via
+`single_task` (i.e. without any iteration space).
+|====
+
+When a function is defined as a free function kernel, each parameter to the
+function is a kernel argument, which must abide by the rules for allowable
+kernel parameter types specified in section 4.12.4 "Rules for parameter passing
+to kernels" of the core SYCL specification.
+The function's return type must be `void`.
+
+The following example demonstrates how a free function kernel using a
+3-dimensional nd-range iteration space can be defined:
+
+```
+SYCL_EXT_ONEAPI_ND_RANGE_KERNEL(3)
+void iota(float start, float *ptr) {
+   // ...
+}
+```
+
+A function decorated with one of these macros can still be called from host
+code.
+The macro has no effect in such cases.
+
+=== New kernel bundle member functions
+
+This extension adds the following new functions which add kernel bundle support
+for free function kernels:
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <auto *Func>
+kernel_id get_kernel_id();
+
+} // namespace sycl::ext::oneapi::experimental
+----
+!====
+
+_Preconditions_: The address `Func` must be the address of some free function
+kernel that is defined in the calling application.
```

Comment:
Should we specify which exception is thrown when `Func` is not an address of a free function kernel?

The same applies to other APIs defined in the extension. I see that it is documented for `ext_oneapi_get_kernel` - we probably just need to copy that "Throws" paragraph everywhere else

Reply:
I intentionally did not specify a "Throws" paragraph on these functions.  The current wording of "Precondition" simply means it is UB if the application passes a `Func` that is not a kernel.

I left this open until we resolve the issue "Can the front-end define a trait like [`is_kernel_function_v`]".  If we can define that trait, then I will change this paragraph from "Precondition" to "Constraint", which means that the compiler must diagnose a compile-time error when `Func` is not the address of a kernel.

In fact, this is the direction I'm leaning.  For an implementation that uses an integration header, I think the header can define the trait.  (There will be some additional restrictions on the kernel function in that case.  See my [comment above](https://github.com/intel/llvm/pull/11861#discussion_r1422522764).)

Reply:
This is addressed in ae6f7ede568821463bed0744594db5a442289f1d.  Now that we have a trait `is_kernel`, we can use that trait as a constraint on these functions.  Thus, the user will get a compile-time error if `Func` is not a kernel function.